vshard/replicaset.lua

```diff
@@ -170,7 +170,9 @@ end
 --
 local function replicaset_connect_to_replica(replicaset, replica)
     local conn = replica.conn
-    if not conn or conn.state == 'closed' then
+    -- gh-341: additionally check if connection's fiber was explicitly killed
+    if not conn or conn.state == 'closed' or (conn.state == 'error_reconnect'
+       and conn._fiber and conn._fiber:status() == 'dead') then
```

Comment:
Are there always such thing as conn._fiber?
```lua
tarantool> c
---
- peer_protocol_version: 3
  peer_uuid: a4c86733-c191-49f1-a2a4-1fa916306e3c
  peer_protocol_features:
    transactions: true
    watchers: true
    error_extension: true
    streams: true
  schema_version: 78
  protocol: Binary
  state: active
  peer_version_id: 133632
  port: '3301'
...

tarantool> c._fiber
---
- null
...

tarantool> _TARANTOOL
---
- 2.10.0-3-gea3ae031e-r493
...
```

In case if fiber is killed connection state is changed to "error":
```lua
tarantool> fiber.find(118)
---
- status: suspended
  name: :3301 (net.box)
  id: 118
...

tarantool> f = fiber.find(118)
---
...

tarantool> f:cancel()
---
...

tarantool> f
---
- the fiber is dead
...

tarantool> c
---
- peer_protocol_version: 3
  peer_uuid: a4c86733-c191-49f1-a2a4-1fa916306e3c
  peer_protocol_features:
    transactions: true
    watchers: true
    error_extension: true
    streams: true
  schema_version: 78
  protocol: Binary
  state: error
  error: fiber is cancelled
  peer_version_id: 133632
  port: '3301'
...
```

Reply:
Yep, there's not always _fiber in connection. I've already fixed that. Speaking of state, as we open connection with `reconnect_after` option the state will be `error_reconnect`. 

Test doesn't fail locally. I'll investigate that thing. The PR is definitely not ready for review.
Sorry for that

Reply:
I'd say that it's a bit strange to rely on netbox internals. It's not documented at could be broken at the any moment (e.g. when https://github.com/tarantool/tarantool/issues/7291 mentioned in https://github.com/tarantool/tarantool/commit/0d944f901cf5400d5832a5528780bb3ecd03ae4b will be fixed).

So but as I showed above nebox with killed worker fiber has
```
  state: error
  error: fiber is cancelled
```

And probably this information could be used. It works for Tarantool 2.10, but for earlier versions it should be checked.

Reply:
> as we open connection with `reconnect_after` option the state will be `error_reconnect`

I was wrong. This is only in the latest releases. E.g. in tarantool 2.6:
```lua
tarantool> c = require('net.box').connect('127.0.0.1:3301', {reconnect_after = 0.1})
---
...

tarantool> c
---
- peer_uuid: 5f56e0a5-1a38-4add-a63b-c139596a2a20
  opts:
    reconnect_after: 0.1
  host: 127.0.0.1
  schema_version: 78
  protocol: Binary
  state: active
  peer_version_id: 132608
  port: '3301'
...

-- <...>: fiber kill

tarantool> c
---
- peer_uuid: 5f56e0a5-1a38-4add-a63b-c139596a2a20
  opts:
    reconnect_after: 0.1
  host: 127.0.0.1
  schema_version: 78
  protocol: Binary
  state: error
  error: fiber is cancelled
  peer_version_id: 132608
  port: '3301'
...

tarantool> _TARANTOOL
---
- 2.6.0-0-g47aa4e01e
...
```

So, we should also check if conn:state() equals to `error`, not only `error_reconnect`.